core/src/main/java/io/opencensus/tags/TagMap.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.tags;
+
+import com.google.common.base.Preconditions;
+import com.google.instrumentation.internal.StringUtil;
+import io.opencensus.tags.TagKey.TagType;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A map from keys to values that can be used to label anything that is associated with a specific
+ * operation.
+ *
+ * <p>For example, {@code TagMap}s can be used to label stats, log messages, or debugging
+ * information.
+ */
+@Immutable
+public final class TagMap {
+  /** The maximum length for a string tag value. */
+  public static final int MAX_STRING_LENGTH = StringUtil.MAX_LENGTH;
+
+  // The types of the TagKey and value must match for each entry.
+  private final Map<TagKey<?>, Object> tags;
+
+  TagMap(Map<TagKey<?>, Object> tags) {
+    this.tags = tags;
+  }
+
+  Map<TagKey<?>, Object> getTags() {
+    return tags;
+  }
+
+  /**
+   * Returns a builder based on this {@code TagMap}.
+   *
+   * @return a builder based on this {@code TagMap}.
+   */
+  public Builder toBuilder() {
+    return new Builder(getTags());
+  }
+
+  /** Builder for the {@link TagMap} class. */
+  public static final class Builder {
```

Comment:
At one point I thought that we will support something like:

public static TagMap build(TagMap parent, List<TagMutation>);

I think this way we can avoid 1 of the 2 copies of the tags.

Maybe not in this PR because we want to move forward but might worth playing with that idea.